// lucy-rest-api-client/Gruntfile.js
/*
 * rest-api-client
 * http://www.bbrennan.info
 *
 * Copyright (c) 2014 Bobby Brennan
 * Licensed under the MIT license.
 */

'use strict';

module.exports = function (grunt) {
  // load all npm grunt tasks
  require('load-grunt-tasks')(grunt);

  // Project configuration.
  grunt.initConfig({
    jshint: {
      all: [
        'Gruntfile.js',
        'tasks/*.js',
        '<%= nodeunit.tests %>'
      ],
      options: {
        jshintrc: '.jshintrc',
        reporter: require('jshint-stylish')
      }
    },

    // Before generating any new files, remove any previously-created files.
    clean: {
      tests: ['tmp']
    },

    // Configuration to be run (and then tested).
    rest_api_client: {
        default_options: {
          language: 'node',
          file: 'github-client.js',
          host: 'api.github.com',
          protocol: 'https',
          headers: {'User-Agent': 'request'},
          endpoints: [{
            name: 'get repos for user',
            path: '/users/:username/repos',
            method: 'get',
            inputs: [{
              label: 'username',
              method: 'path',
            }, {
              label: 'sort',
              default: 'updated',
              constant: true,
              method: 'query',
            }],
            output: {
              type: 'json',
              structure: [{
                name: 'string',
                owner: {
                  login: 'string',
                },
                private: 'boolean',
              }]
            },
          }, {
            name: 'create repo for user',
            path: '/user/repos',
            method: 'post',
            headers: {
              Authorization: 'token :auth_token',
            },
            inputs: [{
              label: 'auth_token',
              method: 'header',
              headerKey: 'Authorization'
            }, {
              label: 'name',
              method: 'json',
              type: 'string',
            }, {
              label: 'description',
              method: 'json',
              type: 'string',
            }, {
              label: 'private',
              method: 'json',
              type: 'boolean',
            }, {
              label: 'auto_init',
              method: 'json',
              type: 'boolean',
              default: true,
            }],
            output: {
              type: 'json',
              structure: {},
            }
          }]
        },
    },

    // Unit tests.
    nodeunit: {
      tests: []//['test/*_test.js']
    }

  });

  // Actually load this plugin's task(s).
  grunt.loadTasks('tasks');

  // Whenever the "test" task is run, first clean the "tmp" dir, then run this
  // plugin's task(s), then test the result.
  grunt.registerTask('test', ['clean', 'rest_api_client', 'nodeunit']);

  // By default, lint and run all tests.
  grunt.registerTask('default', ['test']);

};
